api/external_auth.proto

```diff
@@ -0,0 +1,151 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // The name of the tenant for multi-tenant support.
+  string tenant_name = 1;
+
+  // The request attributes.
+  AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+  // Status OK allows the request. Any other response indicates the request should be denied.
+  google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
+
+    // The canonical service name of the peer.
```

Comment:
What does this mean? Is this the service name as extracted from the `x-envoy-downstream-service-cluster` header? Maybe link back to the relevant docs.

Reply:
We design this API as general precondition API, not specific to Envoy or any system. In a microservice environment, the owner can name their service anyway they like. This API doesn't care about the naming at all.

For example, in mTLS environment, the secure naming system provides the service name. We would not trust HTTP header at all. For MySQL over mTLS, there is no HTTP header at all, but this API design works just fine. How caller resolves the service name is not important to this API design.

Reply:
How is the canonical name of the service different from the principal? Do we need both?

Reply:
Yes. We need both for support secure naming, e.g. the service and the principal running the service. For example, you send a request to api.xyz.com (the service name), and the mTLS says the principal is foo@xyz.com. The authorization system needs to verify that foo@xyz.com is allowed to run api.xyz.com.